src/rules/incognia-authentication.js

```diff
@@ -0,0 +1,67 @@
+/**
+ * @title Incognia Authentication Rule
+ * @overview Verify if the device logging in is at a trusted location.
+ * @gallery true
+ * @category marketplace
+ *
+ */
+async function incogniaAuthenticationRule(user, context, callback) {
+  const _ = require('lodash@4.17.19')
+
+  const { IncogniaAPI } = require('@incognia/api@1.0.0')
+  const { Auth0UserUpdateUtilities } = require('@auth0/rule-utilities@0.2.0')
+
+  const { INCOGNIA_CLIENT_ID, INCOGNIA_CLIENT_SECRET } = configuration;
+
+  if (!INCOGNIA_CLIENT_ID || !INCOGNIA_CLIENT_SECRET) {
+    console.log('Missing required configuration. Skipping.');
+    return callback(null, user, context);
+  }
+
+  // For this rule to be used, please set 'incognia_authentication_rule' as 'enabled' in your Auth0
+  // application metadata. This can be done in the advanced settings of your app.
+  const incogniaAuthenticationRule = _.get(context, 'clientMetadata.incognia_authentication_rule');
+  if (!incogniaAuthenticationRule || incogniaAuthenticationRule !== 'enabled') {
+    console.log('Incognia authentication rule is not enabled for this client. Skipping');
+    return callback(null, user, context);
+  }
+
+  const installationId = _.get(context, 'request.query.installation_id');
+  if (!installationId) {
+    console.log('Missing installation_id. Skipping.');
+    return callback(null, user, context);
+  }
+
+  const accountId = _.get(user, 'user_id');
+  if (!accountId) {
+    console.log('Missing user_id. Skipping.');
+    return callback(null, user, context);
+  }
+
+  let incogniaAPI;
+  if (global.incogniaAPI) {
+    incogniaAPI = global.incogniaAPI;
+  } else {
+    incogniaAPI = new IncogniaAPI({
+      clientId: INCOGNIA_CLIENT_ID,
+      clientSecret: INCOGNIA_CLIENT_SECRET
+    });
+    global.incogniaAPI = incogniaAPI;
+  }
+
+  try {
+    const loginAssessment = await incogniaAPI.registerLoginAssessment({
+      installationId: installationId,
+      accountId: accountId
+    });
+    // TODO: should this be in context.idToken instead? This is information about the request,
```

Comment:
RE: Our conversation from before ... Right now, it looks like we're just storing it on the user profile regardless and not doing anything with the login itself. Should there be a condition where the login is blocked? I'm not totally clear what this Rule would accomplish as-is.

Reply:
Sure. Maybe I can explain what the rule does and then we can figure out the best field to fill.

This rule basically compares the current location of the device logging in (the `installation_id` represents the device) to the location behavior of device(s) that frequently use the account (represented by Auth0's `user.user_id`).

So, Incognia's risk assessment isn't about the user, but about the login attempt. It's a contextual information that does not belong to the user profile (hence my doubt about using `context` instead of `user`). 

I think the best place would be something similar to Auth0's `context.riskAssessment` field, but I think this is an internal field filled by Auth0 only, right? Or could we add information to that?

> Should there be a condition where the login is blocked?

We chose not to block the login in the template but to place the assessment somewhere so that the client can use this information along with other things to make that decision. For example, the client may want to trigger MFA if we say the login is risky but to skip it otherwise.

Is that clearer now? Feel free to ping me through Slack if you think it's better!

Reply:
OK, this makes a bit more sense. 

I can check internally but I would expect that `context.riskAssessment` is internal and not something we should be modifying here. 

What does `loginAssessment` consist of? How customers use this is up to you but I wonder if there aren't things we could key off of from that assessment which would, like you said, trigger MFA or even fail the login if the risk was really bad. Configuration keys could be used to this effect for flagging behavior one way or another.

Reply:
The login assessment basically consists of a payload with two main keys: `riskAssessment`, which may be one of (`low_risk`, `high_risk`, `unknown_risk`) and `evidence`, which states which evidence was used to calculate the risk assessment. You can check out the response here: https://us.incognia.com/apis/transaction-api

So, for example, we were thinking that another rule could use this piece of code to skip MFA:

```
// as we're discussing, we're not sure yet this is the best place for the risk assessment.
const assessment = _.get(user, 'app_metadata.incognia.riskAssessment');

if (!assessment) {
  console.log('Incognia: no risk assessment is available. Skipping adaptive MFA logic');
  return callback(null, user, context);
}

// Example condition: skip MFA for low risk logins
let shouldPromptMfa;

switch (assessment) {
  case 'low_risk':
    shouldPromptMfa = false;
    break;
  case 'unknown_risk':
    shouldPromptMfa = true;
    break;
  case 'high_risk':
    shouldPromptMfa = true;
    break;
}
````

Reply:
Alright, that all makes sense. Thank you for talking the time to explain it out 👍 

I'm still a little worried that folks will turn this integration on and think they are protected, regardless of how well we document it. 

At the very least, maybe we can [add that risk assessment to the ID token](https://auth0.com/docs/rules/examples#hello-world)? Then the mobile app getting the identity back can make a decision to, say, redirect for step-up auth if they have that setup?

Reply:
And just a note from the team that maintains our risk assessment connection here ... they are recommending that we stay away from modifying that object as the shape might change. I think saving this to a different property (with documentation) and passing along in the ID token would work best here.

Reply:
Sure thing! I'll be happy to add the `riskAssessment` to the `ID Token` (`low_risk`, `high_risk` or `unknown_risk`). Do you think we should also add the rest of the payload (namely the `evidence` field) somewhere?

Reply:
I would probably stick with just the risk for now. Just note the custom claim limitations:

https://auth0.com/docs/tokens/create-namespaced-custom-claims

Reply:
Done!